versioned/spi/src/main/java/org/projectnessie/versioned/TracingVersionStore.java

```diff
@@ -0,0 +1,334 @@
+/*
+ * Copyright (C) 2020 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.versioned;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+import javax.annotation.Nonnull;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+
+import io.opentracing.Scope;
+import io.opentracing.Tracer;
+import io.opentracing.Tracer.SpanBuilder;
+import io.opentracing.log.Fields;
+import io.opentracing.tag.Tags;
+import io.opentracing.util.GlobalTracer;
+
+/**
+ * A {@link VersionStore} wrapper that publishes tracing information via OpenTracing/OpenTelemetry.
+ *
+ * @param <VALUE> see {@link VersionStore}
+ * @param <METADATA> see {@link VersionStore}
+ */
+public class TracingVersionStore<VALUE, METADATA, VALUE_TYPE extends Enum<VALUE_TYPE>>
+    implements VersionStore<VALUE, METADATA, VALUE_TYPE> {
+
+  private final VersionStore<VALUE, METADATA, VALUE_TYPE> delegate;
+  private final Supplier<Tracer> tracerSupplier;
+
+  /**
+   * Calls {@link #TracingVersionStore(VersionStore, Supplier)} with the registered
+   * {@link GlobalTracer#get() GlobalTracer}.
+   * @param delegate backing/delegate {@link VersionStore}
+   */
+  public TracingVersionStore(VersionStore<VALUE, METADATA, VALUE_TYPE> delegate) {
+    this(delegate, GlobalTracer::get);
+  }
+
+  /**
+   * Constructor taking the backing/delegate {@link VersionStore} and a {@link Supplier} for
+   * the {@link Tracer} to use.
+   * @param delegate backing/delegate {@link VersionStore}
+   * @param tracerSupplier {@link Supplier} for the {@link Tracer} to use
+   */
+  @VisibleForTesting
+  TracingVersionStore(VersionStore<VALUE, METADATA, VALUE_TYPE> delegate, Supplier<Tracer> tracerSupplier) {
+    this.tracerSupplier = tracerSupplier;
+    this.delegate = delegate;
+  }
+
+  @Override
+  @Nonnull
+  public Hash toHash(@Nonnull NamedRef ref) throws ReferenceNotFoundException {
+    return delegate1Ex("VersionStore.toHash", b -> b
+        .withTag("nessie.version-store.operation", "ToHash")
```

Comment:
I thin the tag names should be `static final String` s

Reply:
Done